[docutils]: Add annotations for `docutils.parsers.rst.states`

Harry-Lees opened the pull request:
Linked Issue: #14031

This PR overlaps with #14191, apologies to @donBarbos, I didn't see your PR while working on this. I think this PR is a more fully implemented version of annotations in `docutils.parsers.rst.states`, but #14191 covers significantly more than just this file.

It might be possible to merge the two together, I will check the other PR to ensure that they line up on the parts which are overlapping.

Edit: I merged some changes where I thought #14191's implementation was more complete than this one, I added Co-authored attribution to the relevant commits, @donBarbos if you are unhappy in any way with the attribution, please let me know how I can update it, I'm not too well versed in Git, but I think I can transfer the commit ownership entirely, or remove you from the commit(s) entirely if you're unhappy to have your name associated with them :)

donBarbos commented:
Thank you, I suggest merging my PR first and then we can see what improvements you have made ;)

I will also note that stubs for `docutils` are already fully generated in my draft PR (WIP) and for now I am sorting out the errors and sending separate updates for directories

srittau commented:
I've merged @donBarbos's PR. This now has (as expected) merge conflicts.

Harry-Lees commented:
I rebased the PR and fixed a couple of obvious errors, I'm sure the CI will have some failures as-well :D.

I need to take a more detailed look at some of the differences, I should be able to get round to it later today.

Harry-Lees commented on stubs/docutils/docutils/parsers/rst/states.pyi:
```diff
@@ -177,84 +185,102 @@ class Inliner:
-    indent = invalid_input  # type: ignore[assignment]
+    blank_finish: Incomplete
+
+    def invalid_input(self, match: Any = None, context: Any = None, next_state: Any = None) -> Never: ...
```
I did this a couple of times with the annotations as `Any`, widening them from the existing stubs here.

The implementation of this function is as [follows](https://github.com/docutils/docutils/blob/7d47c0cc47aa82bb6cb3613773f303d385025e71/docutils/docutils/parsers/rst/states.py#L2505-L2508):

```python
def invalid_input(self, match=None, context=None, next_state=None):
    """Not a compound element member. Abort this state machine."""
    self.state_machine.previous_line()  # back up so parent SM can reassess
    raise EOFError
```

where all three of these variables are clearly unused. I'm not sure if there's precedent for this. This is the reason for a _significant_ number of `Any`'s in this PR, where the variable is unused in the concrete implementation. I'm happy to update in these cases.

srittau replied:
There is a type alias `Unused` in `_typeshed` (alias to `object`, not `Any`), which we use in these cases.

Harry-Lees commented on stubs/docutils/docutils/parsers/rst/states.pyi:
```diff
@@ -264,130 +290,150 @@ class SpecializedBody(Body):
-        self, match: Match[str], context: list[str] | None, next_state: str | None
-    ) -> tuple[list[str], str | None, list[str]]: ...
+        self, match: Match[str], context: Any, next_state: _NextState
+    ) -> tuple[list[Any], _NextState, list[Any]]: ...
```
I would appreciate some input in these cases as-well. The implementing function is:

```python
def field_marker(self, match, context, next_state):
    """Field list field."""
    field, blank_finish = self.field(match)
    self.parent += field
    self.blank_finish = blank_finish
    return [], next_state, []
```

I'm not sure how the `[]` should be annotated ideally. This is another very common pattern which I have always annotated as `list[Any]`. I stayed away from `list[Never]` here, because this will throw an error if the array is appended to (which I can only assume it often is)

AA-Turner replied:
```
    Transition methods all return a 3-tuple:

    - A context object, as (potentially) modified by the transition method.
    - The next state name (a return value of ``None`` means no state change).
    - The processing result, a list, which is accumulated by the state
      machine.
```

srittau reviewed (changes requested):
I'm sorry for the late review.

This looks mostly good, but I noticed that some `context` arguments are annotated with `Any`. We generally need comments explaining non-obvious uses of `Any`. But I also noticed that there already is a `_Context` type var. Could this be used? If not, I would recommend renaming that type var (maybe to `_ContextT`) and instead introduce a `_Context` type alias (potentially to `Any`) and add any explanatory comment to the type var definition.
